helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java

```diff
@@ -93,42 +184,126 @@ public static ClusterModel generateClusterModel(ResourceControllerDataProvider d
-    return generateClusterModel(dataProvider, resourceMap, dataProvider.getEnabledLiveInstances(),
-        Collections.emptyMap(), Collections.emptyMap(), currentStateAssignment);
+  private static Set<AssignableReplica> findToBeAssignedReplicasByComparingBaseline(
+      Map<String, Set<AssignableReplica>> replicaMap, Set<String> activeInstances,
+      Map<String, ResourceAssignment> baselineAssignment,
+      Map<String, ResourceAssignment> bestPossibleAssignment,
+      Map<String, Set<AssignableReplica>> allocatedReplicas) {
+    Set<AssignableReplica> toBeAssignedReplicas = new HashSet<>();
+    // check each resource to identify the allocated replicas and to-be-assigned replicas.
+    for (String resourceName : replicaMap.keySet()) {
+      // <partition, <state, instances set>>
+      Map<String, Map<String, Set<String>>> baselinePartitionStateMap =
+          getValidStateInstanceMap(baselineAssignment.get(resourceName), activeInstances);
+      Map<String, Map<String, Set<String>>> bestPossiblePartitionStateMap =
+          getValidStateInstanceMap(bestPossibleAssignment.get(resourceName), activeInstances);
+      // Iterate the replicas of the resource to find the ones that require reallocating.
+      for (AssignableReplica replica : replicaMap.get(resourceName)) {
+        String partitionName = replica.getPartitionName();
+        String replicaState = replica.getReplicaState();
+        Set<String> baselineAllocations =
+            baselinePartitionStateMap.getOrDefault(partitionName, Collections.emptyMap())
+                .getOrDefault(replicaState, Collections.emptySet());
+        Set<String> bestPossibleAllocations =
+            bestPossiblePartitionStateMap.getOrDefault(partitionName, Collections.emptyMap())
+                .getOrDefault(replicaState, Collections.emptySet());
+
+        // Compare the best possible assignments with the baseline assignment for the common part.
+        List<String> commonAllocations = new ArrayList<>(bestPossibleAllocations);
+        commonAllocations.retainAll(baselineAllocations);
+        if (!commonAllocations.isEmpty()) {
+          // 1. If the partition is allocated at the same location in both baseline and best possible
+          // assignment, there is no need to reassign it.
+          String allocatedInstance = commonAllocations.get(0);
+          allocatedReplicas.computeIfAbsent(allocatedInstance, key -> new HashSet<>()).add(replica);
+          // Remove the instance from the record to prevent the same location being processed again.
+          baselineAllocations.remove(allocatedInstance);
+          bestPossibleAllocations.remove(allocatedInstance);
+        } else if (!baselineAllocations.isEmpty()) {
+          // 2. If the partition is allocated at an active instance in the Baseline but the
+          // allocation does not exist in the best possible assignment, try to rebalance it.
+          toBeAssignedReplicas.add(replica);
```

Comment:
Let's also add what's going on conceptually here:
-----

"This replica exists in the baseline and not in bestPossible. This implies that this replica has been newly added and has not been assigned yet, so it appears in the baseline ("ideal mapping") and does not appear in the bestPossibleAssignment ("reality"). So we add it to `toBeAssignedReplicas` so that it will get assigned.

Reply:
Modified